src/machine/usb_nxpmk66f18.go

```diff
@@ -0,0 +1,626 @@
+				})
+
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_ACM, true, false).Reset()
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_ACM, true, true).Reset()
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_OUT, true, false).Reset()
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_OUT, true, true).Reset()
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_IN, true, false).Reset()
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_IN, true, true).Reset()
+
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_OUT, false, false).Set(make([]byte, usb_BUFFER_SIZE), false)
+				usbBufferDescriptorTable.Get(usb_CDC_ENDPOINT_OUT, false, true).Set(make([]byte, usb_BUFFER_SIZE), true)
+
+				sendZlp()
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (u *USB) stall(n int) {
+	u.ENDPT[n].SetBits(nxp.USB0_ENDPT_EPSTALL)
+}
+
+func (u *USB) unstall(n int) {
+	u.ENDPT[n].ClearBits(nxp.USB0_ENDPT_EPSTALL)
+}
+
+func (u *USB) configureEndpoint(n int, c USBEndpointConfiguration) {
+	var v uint8
+	if c.Transmit {
+		v |= nxp.USB0_ENDPT_EPTXEN
+	}
+	if c.Receive {
+		v |= nxp.USB0_ENDPT_EPRXEN
+	}
+	if c.Handshake {
+		v |= nxp.USB0_ENDPT_EPHSHK
+	}
+	if c.DisableSetup {
+		v |= nxp.USB0_ENDPT_EPCTLDIS
+	}
+	u.ENDPT[n].Set(v)
+}
+
+var (
+	usb0_configuration          volatile.Register8
+	usb0_pending_address_change *uint8
+
+	//go:align 4
+	ep0_rx0_buf [usb_BUFFER_SIZE]byte
+
+	//go:align 4
+	ep0_rx1_buf [usb_BUFFER_SIZE]byte
+
+	ep0_tx_remainder   []byte
+	ep0_tx_bdt_bank    bool
+	ep0_tx_data_toggle bool
+)
+
+func sendUSB0EP0Chunk(data []byte) (remainder []byte) {
+	if len(data) > usb_EP0_SIZE {
+		data, remainder = data[:usb_EP0_SIZE], data[usb_EP0_SIZE:]
+	}
+
+	USB0.debugf("ep0 transmit %d byte(s): %x", len(data), data)
+
+	bd := usbBufferDescriptorTable.Get(0, true, ep0_tx_bdt_bank)
+	bd.Set(data, ep0_tx_data_toggle)
+	ep0_tx_data_toggle = !ep0_tx_data_toggle
+	ep0_tx_bdt_bank = !ep0_tx_bdt_bank
+	return remainder
+}
+
+func sendUSB0EP0(data []byte) {
+	// write to first bank
+	data = sendUSB0EP0Chunk(data)
+	if len(data) == 0 {
+		return
+	}
+
+	// write to second bank
+	data = sendUSB0EP0Chunk(data)
+	if len(data) == 0 {
+		return
+	}
+
+	// save remainder
+	ep0_tx_remainder = data
+}
+
+func sendZlp() {
+	sendUSB0EP0(nil)
+}
+
+func sendUSBPacket(ep uint16, data []byte) {
+	if ep != 0 {
+		panic("not ready to do that")
+	}
+
+	sendUSB0EP0(data)
+}
+
+func (u *USB) CDC() USBCDC {
+	return USBCDC{&u.CDCRXBuffer}
+}
+
+type USBCDC struct {
+	Buffer *RingBuffer
+}
+
+func (u USBCDC) WriteByte(byte) {
+	panic("not implemented")
+}
+
+const (
+	usb_MaxEndpointCount = 16
+	usb_MaxBDTSize       = usb_MaxEndpointCount * 2 * 2
+
+	usb_BDT_Own   = 0x80
+	usb_BDT_Data1 = 0x40
+	usb_BDT_Keep  = 0x20
+	nsb_BDT_NoInc = 0x10
+	usb_BDT_DTS   = 0x08
+	usb_BDT_Stall = 0x04
+
+	usb_BDT_TOK_PID_Mask = 0b00111100
+	usb_BDT_TOK_PID_Pos  = 2
+
+	usb_TOK_PID_OUT   = 0x1
+	usb_TOK_PID_IN    = 0x9
+	usb_TOK_PID_SETUP = 0xD
+)
+
+type register8Aligned32 struct {
```

Comment:
Just to be clear, this struct is not aligned on a 32-bit (4 byte) boundary. It is still 8 bit aligned but simply takes up four bytes. Was this intentional?